src/modules/fw_pos_control/performance_model/PerformanceModel.cpp

```diff
@@ -0,0 +1,225 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+/**
+ * @file PerformanceModel.cpp
+ * Performance model.
+ */
+
+#include <geo/geo.h>
+#include <px4_platform_common/events.h>
+#include "PerformanceModel.h"
+
+// air density of standard athmosphere at 5000m above mean sea level [kg/m^3]
+static constexpr float kAirDensityStandardAtmos5000Amsl = 0.7363f;
+
+// air density of standard athmosphere at 1000m above mean sea level [kg/m^3]
+static constexpr float kAirDensityStandardAtmos1000Amsl = 1.112f;
+
+// [.] minimum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMinWeightRatio = 0.5f;
+
+// [.] maximum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMaxWeightRatio = 2.0f;
+
+// climbrate defining the service ceiling, used to compensate max climbrate based on air density
+static constexpr float kClimbrateMin = 0.5f; // [m/s]
+
+PerformanceModel::PerformanceModel(): ModuleParams(nullptr)
+{
+	updateParams();
+}
+float PerformanceModel::getWeightRatio() const
+{
+	float weight_factor = 1.0f;
+
+	if (_param_weight_base.get() > FLT_EPSILON && _param_weight_gross.get() > FLT_EPSILON) {
+		weight_factor = math::constrain(_param_weight_gross.get() / _param_weight_base.get(), kMinWeightRatio,
+						kMaxWeightRatio);
+	}
+
+	return weight_factor;
+}
+float PerformanceModel::getMaximumClimbRate(float air_density) const
+{
+
+	float climbrate_max = _param_fw_t_clmb_max.get();
+
+	const float density_min = _param_density_min.get();
+
+	if (density_min < kAirDensityStandardAtmos1000Amsl
+	    && density_min > kAirDensityStandardAtmos5000Amsl) {
+		const float density_gradient = math::max((_param_fw_t_clmb_max.get() - kClimbrateMin) /
+					       (CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C -
+						density_min), 0.0f);
+		const float delta_rho = air_density - CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C;
+		climbrate_max = math::max(_param_fw_t_clmb_max.get() + density_gradient * delta_rho, kClimbrateMin);
+	}
+
+	return climbrate_max / getWeightRatio();
+}
+float PerformanceModel::getTrimThrottle(float throttle_min, float throttle_max, float airspeed_sp,
+					float air_density) const
+{
+	const float throttle_trim = getTrimThrottleForAirspeed(airspeed_sp) * getAirDensityThrottleScale(
+					    air_density) * getWeightThrottleScale();
+	return math::constrain(throttle_trim, throttle_min, throttle_max);
+}
+float PerformanceModel::getWeightThrottleScale() const
+{
+	return sqrtf(getWeightRatio());
+}
+
+float PerformanceModel::getAirDensityThrottleScale(float air_density) const
+{
+	float air_density_throttle_scale = 1.0f;
+
+	if (PX4_ISFINITE(air_density)) {
+		// scale throttle as a function of sqrt(rho0/rho)
+		const float eas2tas = sqrtf(CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C / air_density);
```

Comment:
Can't you just constrain the air_density here to kAirDensityStandardAtmos5000Amsl?

Reply:
@RomanBapst this is still unresolved.